Add uv as develop backend command

dmatos2012 opened the pull request:
Hi, 
This should close #1959. However, I would like some feedback/help before I imagine this is ready to merge. I plan on adding more tests to see that it works accross crates etc,  but got into a weird test error loop, that I hope someone can help me here.

Basically about making the `develop_hello_world_uv` test pass. If I activate my `venv` then test fails with `no-such file or directory hello-world`. If I dont activate it, then it passes. Sadly, if I `deactivate` then the other tests dont pass as I need `cffi` to have my tests pass, what gives? 

Doing a `/target/debug/maturin develop --uv` on 3-4 crates from `test-crates` worked fine, thus not sure whats with the test harness that doenst make it happy. 

If anyone else more experienced can help me out, it would be great. Maybe not the best place for long comment here but didnt know where else :). Thanks for your time

konstin commented on src/develop.rs:
```diff
@@ -16,6 +16,10 @@ use std::process::Command;
 use tempfile::TempDir;
 use url::Url;
 
+enum Backend {
```
nit: This should have a more specific name, like `InstallBackend`

konstin commented on src/develop.rs:
```diff
@@ -57,20 +61,30 @@ pub struct DevelopOptions {
     /// `cargo rustc` options
     #[command(flatten)]
     pub cargo_options: CargoOptions,
+    /// Use `uv` as develop backend instead of `pip`
```
```suggestion
    /// Use `uv` to install packages instead of `pip`
```

konstin commented on src/develop.rs:
```diff
@@ -57,20 +61,30 @@ pub struct DevelopOptions {
-            cmd
-        }
-        None => {
+fn make_backend_command(python_path: &Path, pip_path: Option<&Path>, backend: &Backend) -> Command {
```
This should be a method on `Backend`

konstin commented on src/develop.rs:
```diff
@@ -16,6 +16,10 @@ use std::process::Command;
 use url::Url;
 
+enum Backend {
+    Pip,
```
Can you make this `Pip { path: Option<PathBuf> }`?

konstin commented on src/develop.rs:
```diff
@@ -148,12 +164,15 @@ fn pip_install_wheel(
-            &pip_cmd.get_args().collect::<Vec<_>>(),
-            String::from_utf8_lossy(&output.stderr).trim(),
-        );
+        if let Backend::Pip = backend {
```
You can `#[derive(PartialEq, Eq)]` on `Backend` and use the normal `==`

konstin reviewed (commented):
Thanks for starting this!

konstin commented:
> Basically about making the develop_hello_world_uv test pass. If I activate my venv then test fails with no-such file or directory hello-world. If I dont activate it, then it passes.

Can you check where this hello-world comes from with `which hello-world`? The tests should be independent of the containing venv not counting cffi.

dmatos2012 commented:
So,
```rust
# develop_hello_world_uv
#[test]
fn develop_hello_world_uv() {
    handle_result(develop::test_develop(
        "test-crates/hello-world",
        None,
        "develop-hello-world-uv",
        false,
        true,
    ));
}

```
```zsh
> virtualenv venv
created virtual environment CPython3.8.10.final.0-64 in 102ms
  creator CPython3Posix(dest=/home/david/oss/maturin/venv, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/david/.local/share/virtualenv)
    added seed packages: pip==24.0, setuptools==69.1.1, wheel==0.42.0
```
```zsh
> cargo test develop_hello_world_uv (no venv activated)
running 1 test
test develop_hello_world_uv ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 56 filtered out; finished in 3.48s
```
```zsh
> source venv/bin/activate
> pip install cffi

cargo test develop_hello_world_uv
Errors with: 

---- develop_hello_world_uv stdout ----
🔗 Found bin bindings
📡 Using build options bindings from pyproject.toml
🎯 Found 2 Cargo targets in `Cargo.toml`: hello-world, foo
📦 Built wheel to /tmp/.tmpgHeVhr/hello_world-0.1.0-py3-none-linux_x86_64.whl
✏️  Setting installed package as editable
🛠 Installed hello-world-0.1.0
Cause: Check install fail: exit status: 1 
--- Stdout:

--- Stderr:
Traceback (most recent call last):
  File "test-crates/hello-world/check_installed/check_installed.py", line 16, in <module>
    main()
  File "test-crates/hello-world/check_installed/check_installed.py", line 5, in main
    output = check_output(["hello-world"]).decode("utf-8").strip()
  File "/usr/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line 1704, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'hello-world'

thread 'develop_hello_world_uv' panicked at tests/common/mod.rs:88:13:
Check install fail: exit status: 1 
--- Stdout:

--- Stderr:
Traceback (most recent call last):
  File "test-crates/hello-world/check_installed/check_installed.py", line 16, in <module>
    main()
  File "test-crates/hello-world/check_installed/check_installed.py", line 5, in main
    output = check_output(["hello-world"]).decode("utf-8").strip()
  File "/usr/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line 1704, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'hello-world'

note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    develop_hello_world_uv

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 56 filtered out; finished in 7.91s

error: test failed, to rerun pass `--test run`
```

Then, with the `venv` activated and then after the test failure I do:
```zsh
> pip list
Package     Version
----------- -------
cffi        1.16.0
hello-world 0.1.0
pip         24.0
pycparser   2.22
setuptools  69.1.1
wheel       0.42.0
```
```zsh
> hello-world  
Hello, world!

> which hello-world
/home/david/oss/maturin/venv/bin/hello-world
```

But if I change `test_develop_uv` to just use pip , by setting `false` in last arg and recreate then the whole `venv`
```rust
#[test]
fn develop_hello_world_uv() {
    handle_result(develop::test_develop(
        "test-crates/hello-world",
        None,
        "develop-hello-world-uv",
        false,
        false,
    ));
}

```
then 
```zsh
> pip list
Package    Version
---------- -------
cffi       1.16.0
pip        24.0
pycparser  2.22
setuptools 69.1.1
wheel      0.42.0

```
Doesnt show `hello-world` installed
So it looks like when using `uv`, the `hello-world` is installed to the `venv` but when just using `pip` is not?

dmatos2012 commented:
I have also updated the code to reflect the changes @konstin . Thanks for your review! :)

messense commented on src/develop.rs:
```diff
@@ -16,6 +16,42 @@ use std::process::Command;
+
+#[derive(PartialEq, Eq)]
+enum InstallBackend {
+    Pip(Pip),
```
I think this can be simplified to just `Pip { path: Option<PathBuf> }`?

messense commented on tests/common/develop.rs:
```diff
@@ -12,6 +12,7 @@ pub fn test_develop(
     bindings: Option<String>,
     unique_name: &str,
     conda: bool,
+    uv: bool,
```
Might be clearer to also use a enum instead, like

```rust
enum TestInstallBackend {
    Pip,
    Uv,
}
```

messense commented:
> If I activate my `venv` then test fails with `no-such file or directory hello-world`. If I dont activate it, then it passes. Sadly, if I `deactivate` then the other tests dont pass as I need `cffi` to have my tests pass, what gives?

This might be an existing issue, I do remember encountering similar issue in that past. Usually I just install `cffi` globally to fix it.

dmatos2012 commented:
Yeah, addressed your comments @messense. Yeah, and I guess just installing `cffi` globally fixed all my issues and run the tests without a `venv`.  

Finally, before I add the `TestInstallBackend::Uv` for all tests, would you folks be ok with adding `rstest` as a `dev-dependency` crate? That way we can add both `backend` cases without duplicating the test. Let me know, and then ill change/add accordingly to cover all the tests that `TestInstallBackend::Pip` covers

messense commented:
Sounds reasonable to me.

konstin commented:
rstest is a great addition!

For the test suite performance, i'd keep only running the develop test with uv, maybe we can eventually even switch the other tests to uv.

messense reviewed (approved):
LGTM, we should probably also support `uv` other installations not from pypi, but it doesn't need to block merging this.

messense commented:
I'm a bit confused why CI passes without `uv` installed.

messense commented:
I missed that, thanks for pointing it out!
